sequencer/src/block2.rs

```diff
@@ -58,13 +66,168 @@ impl BlockPayload {
+
+        // tx_table[i] is end index for the ith tx,
+        // so the range for the ith tx is tx_table[i-1..i].
+        // But tx_table starts at index 1 in the payload
+        // because index 0 in the payload is the length of the tx_table.
+        // So the range for the ith tx in the payload is actually
+        // [index..index+1].
+        let start = if index == 0 {
+            0
+        } else {
+            self.get_value_usize(index)?
+        }
+        .checked_add(tx_bodies_offset)?;
+
+        let end = self
+            .get_value_usize(index.checked_add(1)?)?
+            .checked_add(tx_bodies_offset)?;
+
+        Some(start..end)
+    }
+
+    // Viewing `self.payload` bytes as a vec of words of type `TxIndex`,
+    // return the `index`th word.
+    fn get_value(&self, index: TxIndex) -> Option<TxIndex> {
+        // TODO idiomatic usize <-> TxIndex conversion.
+        let start = usize::try_from(index)
+            .unwrap()
+            .checked_mul(size_of::<TxIndex>())?;
+
+        let end = start.checked_add(size_of::<TxIndex>())?;
+
+        Some(TxIndex::from_le_bytes(
+            self.payload.get(start..end)?.try_into().unwrap(),
+        ))
+    }
+    fn get_value_usize(&self, index: TxIndex) -> Option<usize> {
+        Some(usize::try_from(self.get_value(index)?).unwrap())
+    }
+
+    // Return length of the tx table
+    // == number of txs in the payload
+    // == the first `TxIndex` word of `self.payload`.
+    fn tx_table_len(&self) -> Option<TxIndex> {
+        self.get_value(0)
+    }
+    fn tx_table_len_usize(&self) -> Option<usize> {
+        self.get_value_usize(0)
+    }
+
+    // Return the index in `self.payload` of the start of the tx bodies (after the tx table).
+    fn tx_bodies_offset(&self) -> Option<usize> {
+        self.tx_table_len_usize()?
+            .checked_add(1)?
+            .checked_mul(size_of::<TxIndex>())
+    }
+}
+
+type TxIndex = <BlockPayload as QueryableBlock>::TransactionIndex;
+
+// TODO upstream type aliases: https://github.com/EspressoSystems/jellyfish/issues/423
+type TxInclusionProof =
+    SmallRangeProof<<UnivariateKzgPCS<Bls12_381> as PolynomialCommitmentScheme>::Proof>;
+
+impl QueryableBlock for BlockPayload {
+    type TransactionIndex = u32;
+    type Iter<'a> = Range<Self::TransactionIndex>;
+    type InclusionProof = TxInclusionProof;
+
+    fn len(&self) -> usize {
+        self.tx_table_len_usize().unwrap_or(0)
+    }
+
+    fn iter(&self) -> Self::Iter<'_> {
+        0..self.tx_table_len().unwrap_or(0)
+    }
+
+    fn transaction_with_proof(
+        &self,
+        index: &Self::TransactionIndex,
+    ) -> Option<(Self::Transaction, Self::InclusionProof)> {
+        let vid = boilerplate::test_vid_factory(); // TODO temporary VID construction
+        let tx_range = self.get_tx_range(*index)?;
+        let proof: SmallRangeProof<_> = vid.payload_proof(&self.payload, tx_range.clone()).unwrap();
```

Comment:
Under what conditions does `payload_proof` fail? Is it ok to panic here or should we return `None`?

Reply:
Whoa thanks for catching this! Fixed in 4e73ce7d79fddb4b06e32ef2d8c533711b08ba24